pulsar-proxy/src/main/java/org/apache/pulsar/proxy/server/URLRegexLookupProxyHandler.java

```diff
@@ -0,0 +1,105 @@
+import org.apache.pulsar.common.api.proto.ServerError;
+import org.apache.pulsar.common.protocol.Commands;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class URLRegexLookupProxyHandler extends DefaultLookupProxyHandler {
+
+    private static final Logger log = LoggerFactory.getLogger(URLRegexLookupProxyHandler.class);
+
+    private String regex;
+
+    private String replacement;
+
+    @Override
+    public void initialize(ProxyService proxy, ProxyConnection proxyConnection) {
+        super.initialize(proxy, proxyConnection);
+        this.regex = proxy.getConfiguration().getProperties().getProperty("urlRegexLookupProxyHandlerRegex");
+        this.replacement =
+            proxy.getConfiguration().getProperties().getProperty("urlRegexLookupProxyHandlerReplacement");
+    }
+
+    @Override
+    public void handleLookup(CommandLookupTopic lookup) {
+        if (log.isDebugEnabled()) {
+            log.debug("Received Lookup from {}", clientAddress);
+        }
+        long clientRequestId = lookup.getRequestId();
+        if (lookupRequestSemaphore.tryAcquire()) {
+            try {
+                LOOKUP_REQUESTS.inc();
+                String serviceUrl = getBrokerServiceUrl(clientRequestId);
+                if (serviceUrl != null) {
+                    if (lookup.isAuthoritative()) {
+                        performLookup(clientRequestId, lookup.getTopic(), serviceUrl, false, 10)
+                            .whenComplete(
+                                (brokerUrl, ex) -> {
+                                    if (ex != null) {
+                                        ServerError serverError = ex instanceof LookupException
+                                            ? ((LookupException) ex).getServerError()
+                                            : getServerError(ex);
+                                        proxyConnection.ctx().writeAndFlush(
+                                            Commands.newLookupErrorResponse(serverError, ex.getMessage(),
+                                                clientRequestId));
+                                    } else {
+                                        proxyConnection.ctx().writeAndFlush(
+                                            Commands.newLookupResponse(brokerUrl, brokerUrl, true,
+                                                CommandLookupTopicResponse.LookupType.Connect, clientRequestId,
+                                                true /* this is coming from proxy */));
+                                    }
+                                });
+                    } else {
+                        performLookup(clientRequestId, lookup.getTopic(), serviceUrl, false, 10)
+                            .whenComplete(
+                                (brokerUrl, ex) -> {
+                                    if (brokerUrl.matches(regex)) {
+                                        String proxyUrl = brokerUrl.replaceAll(regex, replacement);
```

Comment:
could we optimize here if the proxyUrl is *this* proxy? we could go directly to the broker in this case

Reply:
I'm not sure a proxy knows its URL